### PR TITLE
support PIC code very limited and some bug fix

### PR DESCRIPTION
**Pull request by ParkHanbum**
Summary
- insn_check.size did not setted before using it, so fix it.
- change jumping to unknown address treat as normal.

check_instrumentable function update as,
- print more detail information when instruction could not supported.
- return specific reason when instruction could not supported. ( this can be one or more)

- manage values to be used in full-dynamic process in one structure.
- limited support for PIC codes which has 'LEA' opcode.

### CODE REVIEW

File: arch/x86_64/mcount-insn.c [hunk 1]
```diff
@@ -31,66 +31,102 @@ void mcount_disasm_finish(struct mcount_disasm_engine *disasm)
...
+	int status = 0;
 
 	/*
 	 * 'detail' can be NULL on "data" instruction
 	 * if SKIPDATA option is turned ON
 	 */
-	if (insn->detail == NULL)
-		return false;
+	if (insn->detail == NULL) {
+		status = INSTRUMENT_FAIL_UNKNOWN;
```
Comment (namhyung) [inline arch/x86_64/mcount-insn.c]: It could be `_NODETAIL`.

File: arch/x86_64/mcount-insn.c [hunk 2]
```diff
@@ -31,66 +31,102 @@ void mcount_disasm_finish(struct mcount_disasm_engine *disasm)
...
+
+			case X86_OP_IMM:
+				if (jmp_or_call > 0) {
+					if (jmp_or_call == OP_GROUP_CALL)
+						status |= INSTRUMENT_FAIL_RELCALL;
+					else
+						status |= INSTRUMENT_FAIL_RELJMP;
+					goto out;
+				}
+				continue;
```
Comment (namhyung) [inline arch/x86_64/mcount-insn.c]: Please consider reducing indent levels by bailing out the short case.

```c
case X86_OP_IMM:
        if (!jmp_or_call)
                continue;
        ...
```

File: arch/x86_64/mcount-dynamic.c [hunk 3]
```diff
@@ -441,10 +441,10 @@ bool add_bad_jump(struct mcount_dynamic_info *mdi, unsigned long callsite,
 		return true;
 
 	sym = find_sym(&mdi->map->mod->symtab, target - mdi->map->start);
+	/* only care about jumps to the middle of a function */
 	if (sym == NULL)
-		return true;
+		return false;
```
Comment (namhyung) [inline arch/x86_64/mcount-dynamic.c]: This is just to be conservative since we don't know it's correct or not.  Let's make it robust even if it cannot trace some.

File: arch/x86_64/mcount-insn.c [hunk 4]
```diff
@@ -170,10 +180,8 @@ static bool check_unsupported(struct mcount_disasm_engine *disasm, cs_insn *insn
 
 			/* disallow jump to middle of other function */
 			if (insn_check->addr > target ||
-			    target >= insn_check->addr + insn_check->func_size) {
-				/* also mark the target function as invalid */
+			    target >= insn_check->addr + insn_check->func_size)
```
Comment (namhyung) [inline arch/x86_64/mcount-insn.c]: Why did you remove the comment?

File: arch/x86_64/mcount-dynamic.c [hunk 5]
```diff
@@ -461,18 +461,18 @@ bool add_bad_jump(struct mcount_dynamic_info *mdi, unsigned long callsite,
...
-	uint64_t sym_addr = sym->addr + mdi->map->start;
+	struct mcount_instrument_info *info;
+
+	info = disasm_check_insns(disasm, mdi, sym);
 
-	instr_size = disasm_check_insns(disasm, mdi, sym);
-	if (instr_size < CALL_INSN_SIZE)
-		return instr_size;
+	if (info->size < CALL_INSN_SIZE)
+		return info->size;
```
Comment (namhyung) [inline arch/x86_64/mcount-dynamic.c]: Who's gonna free the info?

File: arch/x86_64/mcount-insn.c [hunk 6]
```diff
@@ -189,24 +197,35 @@ static bool check_unsupported(struct mcount_disasm_engine *disasm, cs_insn *insn
...
 	uint8_t endbr64[] = { 0xf3, 0x0f, 0x1e, 0xfa };
-	int ret = INSTRUMENT_FAILED;
 	struct disasm_check_data insn_check = {
 		.addr		= sym->addr + mdi->map->start,
 		.func_size	= sym->size,
 	};
 	struct dynamic_bad_symbol *badsym;
 
+	struct mcount_instrument_info *info =
+		xmalloc(sizeof(struct mcount_instrument_info));
```
Comment (namhyung) [inline arch/x86_64/mcount-insn.c]: The convention is to use the variable name like `xmalloc(sizeof(*info))` since it's more flexible to change.

Comment (namhyung) [reply arch/x86_64/mcount-insn.c]: And please put a blank line **AFTER** the declaration.

File: arch/x86_64/mcount-insn.c [hunk 7]
```diff
@@ -58,6 +58,245 @@ void print_instrument_fail_msg(int reason)
...
+			break;
+		case X86_REG_RSI:
+			mov_insns[reg_offset] = mov_opnd_48[5];
+			break;
+		case X86_REG_RBP:
+			mov_insns[reg_offset] = mov_opnd_48[6];
+			break;
+		case X86_REG_RSP:
+			mov_insns[reg_offset] = mov_opnd_48[7];
+			break;
```
Comment (namhyung) [inline arch/x86_64/mcount-insn.c]: We don't indent `case` and it seems you can remove duplication here using reg -> index table.

File: arch/x86_64/mcount-insn.c [hunk 8]
```diff
@@ -129,7 +129,7 @@ static bool check_unsupported(struct mcount_disasm_engine *disasm, cs_insn *insn
 
 			/* disallow (back) jump to the prologue */
 			if (insn_check->addr <= target &&
-			    target < insn_check->addr + insn_check->size)
+			    target < insn_check->addr + insn_check->copy_size)
```
Comment (namhyung) [inline arch/x86_64/mcount-insn.c]: Nice catch!

File: arch/x86_64/mcount-insn.c [hunk 9]
```diff
@@ -31,66 +33,349 @@ void mcount_disasm_finish(struct mcount_disasm_engine *disasm)
...
+
+	/*
+	 * array for lea instruction.
+	 * ex) lea rbx, [rbx+0x8f3f85]
+	 */
+	uint8_t lea_insns[] = {
+		0x48, 0x8d, 0x00, 0x00, 0x00, 0x00, 0x00
+	};
+	int lea_rex_offset = 0;
+	int lea_insn_offset = 1;
```
Comment (honggyukim) [inline arch/x86_64/mcount-insn.c]: It's just minor but this variable is not used anywhere.

**Comment by honggyukim** [comment]
Hi @ParkHanbum. I just tested this with chrome, but it has some regression bugs.

The following example works fine on master
```
$ uftrace record --no-libcall --no-event -P. -Z 1259 ./chrome --no-sandbox --headless https://www.google.com
[0909/085731.624179:WARNING:spdy_session.cc(3180)] Received HEADERS for invalid stream 15
```
However, this PR seems to introduce some bugs and it gets crashed as follows:
```
$ uftrace record --no-libcall --no-event -P. -Z 1259 ./chrome --no-sandbox --headless https://www.google.com
WARN: Segmentation fault: address not mapped (addr: (nil))
WARN: Backtrace from uftrace:
WARN: =====================================
WARN: [0] (InitializeSystemInfo[55fb9dc88255] <= NumCPUs[55fb9dc88879])
WARN: child terminated by signal: 11: Segmentation fault
```

File: arch/x86_64/mcount-dynamic.c (hunk 3 above)
Comment (ParkHanbum) [reply arch/x86_64/mcount-dynamic.c]: @namhyung  I think this is reasonable. because we can not patch the function without symbol. so, jumping to not recognized address in symbol no need to exclude.

**Comment by ParkHanbum** [comment]
@honggyukim thank you to testing. would you brief me more details?

**Comment by ParkHanbum** [comment]
@namhyung thanks, and I wonder about testing list for dynamic tacing. I had tested it with nodejs and python only. so, let me know what program do we test for testing dynamic tracing.

**Comment by namhyung** [comment]
I mean it's nice if you run your test workload (node and python) before and after your change.  With -v option it'd show dynamic patch stat, and compare the numbers.

**Comment by ParkHanbum** [comment]
@namhyung here is the test records : 
```
# test arguments for node
# ./uftrace record -v -L. -P. `which node` -e 'console.log("hello")'
#test arguments for python
# $ ./uftrace record -v -L. -P. `which python` -c 'print "Hello"'
```


first : records of master branch.
```
$ git log -1 --pretty=oneline
3ebf01492d3069369961f336ee1ad006ab619daf (HEAD -> master, origin/master, origin/HEAD) Merge branch 'info-update'

dynamic: dynamic patch stats for 'node'
dynamic:    total:    17538
dynamic:  patched:    15039 (85.75%)
dynamic:   failed:     2499 (14.24%)
dynamic:  skipped:        0 ( 0.00%)

dynamic: dynamic patch stats for 'python2.7'
dynamic:    total:     1109
dynamic:  patched:      259 (23.35%)
dynamic:   failed:      834 (75.20%)
dynamic:  skipped:       16 ( 1.44%)
dynamic: no match:        0
```


1st :  jumping to unknown address exclude from bad jump cases
```
$ git log -1 --pretty=oneline
c8cdbbba6442b8558ec67b875424858c387e500f (HEAD) dynamic: jumping to unknown address exclude from bad jump cases
...
dynamic: dynamic patch stats for 'node'
dynamic:    total:    17538
dynamic:  patched:    15262 (87.02%)
dynamic:   failed:     2276 (12.97%)
dynamic:  skipped:        0 ( 0.00%)
dynamic: no match:        0

dynamic: dynamic patch stats for 'python2.7'
dynamic:    total:     1109
dynamic:  patched:      663 (59.78%)
dynamic:   failed:      430 (38.77%)
dynamic:  skipped:       16 ( 1.44%)
dynamic: no match:        0
```

2nd : support position indepdent code if opcode is 'LEA' only
```
$ git log -1 --pretty=oneline
c1705efd865f8104791b8a46a9412c550e88efa5 (HEAD -> enhance-full-dynamic, work/enhance-full-dynamic) dynamic: support position indepdent code if opcode is 'LEA' only

dynamic: dynamic patch stats for 'node'
dynamic:    total:    17538
dynamic:  patched:    16335 (93.14%)
dynamic:   failed:     1203 ( 6.85%)
dynamic:  skipped:        0 ( 0.00%)
dynamic: no match:        0

dynamic: dynamic patch stats for 'python2.7'                                                                                                                                                 dynamic:    total:     1109                                                                                                                                                                  dynamic:  patched:      797 (71.86%)                                                                                                                                                         dynamic:   failed:      296 (26.69%)                                                                                                                                                         dynamic:  skipped:       16 ( 1.44%)                                                                                                                                                         dynamic: no match:        0  

```

**Comment by ParkHanbum** [comment]
@namhyung @honggyukim  I fixed some mistakes. Would you like to review this PR later?

File: arch/x86_64/mcount-insn.c (hunk 4 above)
Comment (ParkHanbum) [reply arch/x86_64/mcount-insn.c]: this is mistake. I recover it.

File: arch/x86_64/mcount-insn.c (hunk 2 above)
Comment (ParkHanbum) [reply arch/x86_64/mcount-insn.c]: thanks for advise

File: arch/x86_64/mcount-insn.c [hunk 10]
```diff
@@ -31,66 +31,102 @@ void mcount_disasm_finish(struct mcount_disasm_engine *disasm)
...
+#define INSTRUMENT_FAIL_RELCALL		0x00000001L << 3
+#define INSTRUMENT_FAIL_PICCODE		0x00000001L << 4
+
+#define OP_GROUP_JMP			1
+#define OP_GROUP_CALL			2
+
+void print_instrument_fail_msg(int reason)
+{
+	if (reason & INSTRUMENT_FAIL_NOOPRND) {
+		pr_dbg3("\tNot supported opcode without operand\n");
```
[the remark this answers is not in the document]
Comment (ParkHanbum) [reply arch/x86_64/mcount-insn.c]: this is mistypo.

File: arch/x86_64/mcount-insn.c [hunk 11]
```diff
@@ -31,66 +31,102 @@ void mcount_disasm_finish(struct mcount_disasm_engine *disasm)
 	cs_close(&disasm->engine);
 }
 
+#define INSTRUMENT_FAIL_UNKNOWN		0x00000001L
+#define INSTRUMENT_FAIL_NOOPRND		0x00000001L << 1
+#define INSTRUMENT_FAIL_RELJMP		0x00000001L << 2
+#define INSTRUMENT_FAIL_RELCALL		0x00000001L << 3
+#define INSTRUMENT_FAIL_PICCODE		0x00000001L << 4
```
[the remark this answers is not in the document]
Comment (ParkHanbum) [reply arch/x86_64/mcount-insn.c]: yes, there can be exist fail reason more than one. if we try to support this, we must know all reasons why it failed.

File: arch/x86_64/mcount-insn.c [hunk 12]
```diff
@@ -31,66 +31,102 @@ void mcount_disasm_finish(struct mcount_disasm_engine *disasm)
...
+
+#define OP_GROUP_JMP			1
+#define OP_GROUP_CALL			2
```
[the remark this answers is not in the document]
Comment (ParkHanbum) [reply arch/x86_64/mcount-insn.c]: thanks for advise. I changed the codes by referring codes of uftrace.

File: arch/x86_64/mcount-insn.c (hunk 1 above)
Comment (ParkHanbum) [reply arch/x86_64/mcount-insn.c]: I agree, it was changed.

File: arch/x86_64/mcount-insn.c (hunk 6 above)
Comment (ParkHanbum) [reply arch/x86_64/mcount-insn.c]: thanks for advise, it was changed.

File: arch/x86_64/mcount-dynamic.c (hunk 5 above)
Comment (ParkHanbum) [reply arch/x86_64/mcount-dynamic.c]: last line of patch_normal_func function.

File: arch/x86_64/mcount-insn.c (hunk 7 above)
Comment (ParkHanbum) [reply arch/x86_64/mcount-insn.c]: moved indent.

**Comment by honggyukim** [comment]
Thanks for the update. But it still has the same problem.
```
$ uftrace record --no-libcall --no-event -P. -Z 1259 ./chrome --no-sandbox --headless https://www.google.com
WARN: process crashed by signal 11: Segmentation fault (si_code: 128)
WARN: Backtrace from uftrace:
WARN: =====================================
WARN: [0] (InitializeSystemInfo[561de546b255] <= NumCPUs[561de546b879])
WARN: child terminated by signal: 11: Segmentation fault
```

**Comment by honggyukim** [comment]
Here are some disassembled code for problematic functions.
```asm
00000000026e9250 <InitializeSystemInfo()>:
 26e9250:       55                      push   %rbp
 26e9251:       48 89 e5                mov    %rsp,%rbp
 26e9254:       41 57                   push   %r15
 26e9256:       41 56                   push   %r14
 26e9258:       41 55                   push   %r13
 26e925a:       41 54                   push   %r12
 26e925c:       53                      push   %rbx
 26e925d:       48 81 ec 38 08 00 00    sub    $0x838,%rsp
 26e9264:       64 48 8b 04 25 28 00    mov    %fs:0x28,%rax
 26e926b:       00 00
 26e926d:       48 89 45 d0             mov    %rax,-0x30(%rbp)
...
```
```asm
00000000026e9870 <NumCPUs()>:
 26e9870:       55                      push   %rbp
 26e9871:       48 89 e5                mov    %rsp,%rbp
 26e9874:       e8 d7 f9 ff ff          callq  26e9250 <InitializeSystemInfo()>
 26e9879:       8b 05 99 b7 7c 06       mov    0x67cb799(%rip),%eax        # 8eb5018 <cpuinfo_num_cpus>
 26e987f:       5d                      pop    %rbp
 26e9880:       c3                      retq
 26e9881:       cc                      int3
 26e9882:       cc                      int3
 26e9883:       cc                      int3
...
```

**Comment by honggyukim** [comment]
After rebasing this branch onto [review/dynamic-unpatch-v2](), I tested it again as follows:
```
$ uftrace record --no-libcall --no-event -P. -Z 1259 -U InitializeSystemInfo ./chrome --no-sandbox --headless https://www.google.com
WARN: Segmentation fault: invalid permission (addr: 0x55d401be2128)
WARN: Backtrace from uftrace:
WARN: =====================================
WARN: [0] (std::__1::ios_base::Init::Init[55d401bcb115] <= _GLOBAL__sub_I_iostream.cpp[55d401bcc2a9])
WARN: child terminated by signal: 11: Segmentation fault
```
It unpatched the problematic function, but crashed by another function `std::__1::ios_base::Init::Init` anyway.
```asm
00000000047fe110 <std::__1::ios_base::Init::Init()>:
 47fe110:       55                      push   %rbp
 47fe111:       48 89 e5                mov    %rsp,%rbp
 47fe114:       41 57                   push   %r15
 47fe116:       41 56                   push   %r14
 47fe118:       41 55                   push   %r13
 47fe11a:       41 54                   push   %r12
 47fe11c:       53                      push   %rbx
 47fe11d:       48 83 ec 28             sub    $0x28,%rsp
 47fe121:       4c 8b 3d f0 df c4 04    mov    0x4c4dff0(%rip),%r15        # 944c118 <_DYNAMIC+0x4470>
 47fe128:       49 8b 1f                mov    (%r15),%rbx
 47fe12b:       4c 8d 35 c6 18 d3 04    lea    0x4d318c6(%rip),%r14        # 952f9f8 <std::__1::__cin>
 47fe132:       4c 89 f7                mov    %r14,%rdi
...
```
I unpatched `std::__1::ios_base::Init::Init` additionally, but it still gets crashed.
```
$ uftrace record --no-libcall --no-event -P. -Z 1259 -U InitializeSystemInfo -U std::__1::ios_base::Init::Init ./chrome --no-sandbox --headless https://www.google.com
WARN: child terminated by signal: 11: Segmentation fault
```

**Comment by namhyung** [comment]
Oh, success rate goes much better, good job!  I'll review later.

File: arch/x86_64/mcount-insn.c [hunk 13]
```diff
@@ -63,6 +63,248 @@ void print_instrument_fail_msg(int reason)
...
+	};
+
+	// REX.4D operands,
+	uint8_t lea_opnd_4D[] = {
+		//r8,	r9,	r10,	r11,	r12,	r13,	r14,	r15
+		0x80,	0x89,	0x92,	0x9b,	0xa4,	0xad,	0xb6,	0xbf
+	};
+
+	// for currently, support LEA instruction only.
+	if (strncmp(insn->mnemonic, "lea", 3) != 0)
```
Comment (namhyung) [inline arch/x86_64/mcount-insn.c]: There's LEAVE instruction too.

File: arch/x86_64/mcount-insn.c [hunk 14]
```diff
@@ -63,6 +63,248 @@ void print_instrument_fail_msg(int reason)
...
+		goto out;
+
+	// according to intel manual, lea instruction takes 2 operand only.
+	cs_x86_op *opnd1 = &(x86->operands[0]);
+	cs_x86_op *opnd2 = &(x86->operands[1]);
```
Comment (namhyung) [inline arch/x86_64/mcount-insn.c]: I think you need to check the second operand is actually RIP.

**Comment by honggyukim** [comment]
Before
```
$ uftrace record --no-libcall --no-event -P. -Z 1259 ./chrome --no-sandbox --headless https://www.google.com
WARN: process crashed by signal 11: Segmentation fault (si_code: 128)
WARN: Backtrace from uftrace:
WARN: =====================================
WARN: [0] (InitializeSystemInfo[55bca52bb255] <= NumCPUs[55bca52bb879])
WARN: child terminated by signal: 11: Segmentation fault
```
After applying the change below:
```diff
diff --git a/libmcount/dynamic.c b/libmcount/dynamic.c
index 849c903..a907848 100644
--- a/libmcount/dynamic.c
+++ b/libmcount/dynamic.c
@@ -90,7 +90,7 @@ struct mcount_orig_insn *mcount_save_code(struct mcount_instrument_info *info,
        orig = lookup_code(&code_tree, info->addr, true);
        orig->insn = cp->page + cp->pos;

-       memcpy(orig->insn, (void *)info->insns, info->insns_size);
+       memcpy(orig->insn, (void *)info->addr, info->insns_size);
        memcpy(orig->insn + info->insns_size, jmp_insn, jmp_size);

        cp->pos += patch_size;
```
It gets crashes with a different error message as follows:
```
$ uftrace record --no-libcall --no-event -P. -Z 1259 ./chrome --no-sandbox --headless https://www.google.com
Received signal 11 SEGV_MAPERR 7f3154de4c38
#0 0x7f314ddc8df7 backtrace
#1 0x561995e4f14f (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x4ca214e)
#2 0x561995e4ecb1 (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x4ca1cb0)
#3 0x7f314d9ae390 <unknown>
#4 0x7f314e2160c7 <unknown>
  r8: 00000000021cbfbb  r9: 00000000000007ff r10: 2eda191ea6b8b25c r11: 00007f31479a8f90
 r12: 0000000000000000 r13: 00003299d24f6210 r14: 000011490b823bcf r15: 00007ffefce75dd0
  di: 0000000000000fff  si: 0000000000000100  bp: 00007ffefce75c80  bx: 000056199a6b7138
  dx: 0000000000000000  ax: 00007f314e217240  cx: 000056199a062200  sp: 00007ffefce75c68
  ip: 00007f314e2160c7 efl: 0000000000010206 cgf: 0000000000000033 erf: 0000000000000004
 trp: 000000000000000e msk: 0000000000000000 cr2: 00007f3154de4c38
[end of stack trace]
Calling _exit(1). Core file will not be generated.
```
I tested and confirmed that it works fine and the previous problem is fixed if the last commit https://github.com/namhyung/uftrace/pull/856/commits/c1705efd865f8104791b8a46a9412c550e88efa5 is reverted.

File: libmcount/dynamic.c [hunk 15]
```diff
@@ -87,11 +87,11 @@ struct mcount_orig_insn *mcount_save_code(unsigned long addr, unsigned insn_size
 		list_add_tail(&cp->list, &code_pages);
 	}
 
-	orig = lookup_code(&code_tree, addr, true);
+	orig = lookup_code(&code_tree, info->addr, true);
 	orig->insn = cp->page + cp->pos;
 
-	memcpy(orig->insn, (void *)addr, insn_size);
-	memcpy(orig->insn + insn_size, jmp_insn, jmp_size);
+	memcpy(orig->insn, (void *)info->insns, info->insns_size);
```
[the remark this answers is not in the document]
Comment (namhyung) [reply libmcount/dynamic.c]: No, this is intended.  The `info->insns` contains original and new instructions which can be added for "LEA"

**Comment by namhyung** [comment]
So I now can run python interpreter which below change:

```diff
diff --git a/arch/x86_64/mcount-insn.c b/arch/x86_64/mcount-insn.c
index 3aaa346..67aa34c 100644
--- a/arch/x86_64/mcount-insn.c
+++ b/arch/x86_64/mcount-insn.c
@@ -75,8 +75,7 @@ void print_instrument_fail_msg(int reason)
  */
 static int handle_pic(cs_insn *insn, uint8_t insns[])
 {
-       cs_x86 *x86;
-       x86 = &insn->detail->x86;
+       cs_x86 *x86 = &insn->detail->x86;
 
        /*
         * array for mov instruction.
@@ -126,13 +125,20 @@ static int handle_pic(cs_insn *insn, uint8_t insns[])
        };
 
        // for currently, support LEA instruction only.
-       if (strncmp(insn->mnemonic, "lea", 3) != 0)
+       if (strcmp(insn->mnemonic, "lea") != 0)
                goto out;
 
        // according to intel manual, lea instruction takes 2 operand only.
        cs_x86_op *opnd1 = &(x86->operands[0]);
        cs_x86_op *opnd2 = &(x86->operands[1]);
 
+       if (opnd2->type != X86_OP_MEM || opnd2->mem.base != X86_REG_RIP)
+               goto out;
+
+       // handling the SIB not supported yet.
+       if (opnd2->mem.scale > 1 || opnd2->mem.disp == 0)
+               goto out;
+
        // get the register from opnd1 to load rip address
        if (X86_REG_RAX <= opnd1->reg && opnd1->reg <= X86_REG_RSP) {
                mov_insns[mov_rex_offset] = 0x48;
@@ -205,10 +211,6 @@ static int handle_pic(cs_insn *insn, uint8_t insns[])
        PIC_base += insn->size;
        *(uint64_t *)(mov_insns + mov_imm64_offset) = PIC_base;
 
-       // handling the SIB not supported yet.
-       if (opnd2->mem.scale > 1 || opnd2->mem.disp == 0)
-               return -1;
-
        switch (opnd1->reg) {
        //rax,  rbx,    rcx,    rdx,    rdi,    rsi,    rbp,    rsp
        case X86_REG_RAX:
@@ -515,7 +517,8 @@ disasm_check_insns(struct mcount_disasm_engine *disasm,
                        info->size = 0;
                }
                else {
-                       info->insns = xrealloc(info->insns, insns_size);
+                       info->insns = xrealloc(info->insns,
+                                              info->insns_size + insns_size);
                }
                memcpy(info->insns + info->insns_size, (void *)insns_byte, insns_size);
                info->insns_size += insns_size;
```

**Comment by namhyung** [comment]
```
$ uftrace -P. -D 3 -t 10us /usr/bin/python -c 'print("hello")'
hello
# DURATION     TID     FUNCTION
            [547914] | Py_Initialize() {
  20.683 us [547914] |   memcpy();
            [547914] |   PyImport_ImportModule() {
  51.620 us [547914] |     _PyObject_GenericGetAttrWithDict();
  72.433 us [547914] |   } /* PyImport_ImportModule */
  77.290 us [547914] |   PyImport_ImportModule();
            [547914] |   PyImport_ImportModule() {
 187.658 us [547914] |     PyMarshal_ReadObjectFromString();
   6.969 ms [547914] |     PyEval_EvalCode();
   7.237 ms [547914] |   } /* PyImport_ImportModule */
  32.456 us [547914] |   setlocale();
  29.280 us [547914] |   PyMarshal_ReadObjectFromString();
            [547914] |   PyEval_EvalCode() {
  10.809 us [547914] |     fread();
 157.657 us [547914] |     PyMarshal_ReadObjectFromString();
 166.958 us [547914] |     PyEval_EvalCode();
  77.531 us [547914] |     PyMarshal_ReadObjectFromString();
 610.944 us [547914] |   } /* PyEval_EvalCode */
  23.199 us [547914] |   PyMarshal_ReadObjectFromString();
  53.728 us [547914] |   PyEval_EvalCode();
  10.621 ms [547914] | } /* Py_Initialize */
            [547914] | PyRun_StringFlags() {
  11.901 us [547914] |   PyParser_ASTFromString();
  13.175 us [547914] |   PyEval_EvalCode();
  38.637 us [547914] | } /* PyRun_StringFlags */
  16.755 us [547914] | PyType_Modified();
 380.956 us [547914] | PyGC_Collect();
```

**Comment by namhyung** [comment]
Although the success rate was still not good enough, it gets 5% increment.

before:
```
dynamic: dynamic patch type: python2.7: 0 (none)
dynamic: dynamic patch stats for 'python2.7'
dynamic:    total:     1109
dynamic:  patched:      260 (23.44%)
dynamic:   failed:      832 (75.02%)
dynamic:  skipped:       17 ( 1.53%)
dynamic: no match:        0
```
after:
```
dynamic: dynamic patch type: python2.7: 0 (none)
dynamic: dynamic patch stats for 'python2.7'
dynamic:    total:     1109
dynamic:  patched:      317 (28.58%)
dynamic:   failed:      775 (69.88%)
dynamic:  skipped:       17 ( 1.53%)
dynamic: no match:        0
```

**Comment by ParkHanbum** [comment]
@namhyung I don't understand why you need to do it? I already tested python 2.7 and it was succeed. what problem do you have with python?

and PIC support do not increase patch status dramatically. but I think you still exclude function that have code which jumping to unknown address. this is highly decrease patching successful rates, as I mention above, only 23.35% success with it. after ignore it, it increase as 59.79%.

**Comment by namhyung** [comment]
I got segfault without the change and it's obvious.

**Comment by ParkHanbum** [comment]
@namhyung could you please share with me about problem that you have fixed? and I wonder that how about removing the exclusion of code that jumps to unknown addresses?

**Comment by namhyung** [comment]
I already left comments above for them:
1. there's "LEAVE" instruction can pass strncmp() check
2. you did realloc() with old size not (old size + insn size)

**Comment by namhyung** [comment]
@ParkHanbum I uploaded [check/dynamic-pic-lea](https://github.com/namhyung/uftrace/compare/check/dynamic-pic-lea) which have two more cleanup patches.  Please take a look!

**Comment by ParkHanbum** [comment]
@namhyung what I say? it is good! I had a plan to clean codes about remove no need codes since added mcount_disasm_info, after merged this PR. but you did it, so thanks to you.

**Comment by honggyukim** [comment]
Unlike https://github.com/namhyung/uftrace/pull/856#issuecomment-529257870, confirmed that it doesn't get crashed.
```
$ uftrace record --no-libcall --no-event -P. -Z 1259 ./chrome --no-sandbox --headless https://www.google.com
```
The trace data looks good.
